subprojects/core/src/main/java/org/gradle/initialization/DefaultGradleLauncherFactory.java

```diff
@@ -147,6 +148,22 @@ private GradleLauncher doNewInstance(
         }
 
         DeprecationLogger.init(usageLocationReporter, startParameter.getWarningMode(), serviceRegistry.get(BuildOperationProgressEventEmitter.class));
+        @SuppressWarnings("deprecation")
+        File customSettingsFile = startParameter.getSettingsFile();
```

Comment:
Why don't we emit the deprecation warning in `StartParameter.setSettingsFile()` or at the call sites of `StartParameter.setSettingsFile()`? Based on that we could offer advice on how to resolve the deprecation warning:
- remove the `-c` argument from the command line
- pinpoint which plugin/etc contributes the custom start parameter

Reply:
See my comment here: https://github.com/gradle/gradle/pull/16713#issuecomment-814005350

At the time those setters are called for the command-line options case, we are on the client-side where we do not have the logging infrastructure initialized. That's why I intercept those options set here in the daemon, just after the logging infrastructure is ready. And at this point, I no longer have the context of when exactly those were set, if it was some custom user code, a GradleBuild task, or a command-line option.
The way I see this problem:
- GradleBuild task can easily emit the deprecation when the deprecated method is called.
- Emitting deprecation in the StartParameter setters would catch GradleBuild+user code origins. In this case the deprecation emitted with GradleBuild would be more specific, but then users would get another one, pointing at StartParameter usage, which can be confusing.
- Lastly, the current place of interception - that catches all three possible usages - if we also have any of the above two, users would get multiple different deprecation messages for the same problem.

Given this, I thought it would be simpler to have a single deprecation emitted here with a pointer to the documentation detailing the three possible causes.

Reply:
After reading through this comment (and the one above), I tend to agree. There's not much extra gain in distinguishing all those cases, but the implementation would be much more complex. Let's keep it as it is.